ros2_control_test_assets/include/ros2_control_test_assets/components_urdfs.hpp

```diff
@@ -332,6 +332,42 @@ const auto valid_urdf_ros2_control_actuator_only =
   </ros2_control>
 )";
 
+// 10. Industrial Robots with integrated GPIO
+const auto valid_urdf_ros2_control_system_robot_with_gpio =
+  R"(
+  <ros2_control name="RRBotSystemWithGPIO" type="system">
+    <hardware>
+      <plugin>ros2_control_demo_hardware/RRBotSystemWithGPIOHardware</plugin>
+      <param name="example_param_write_for_sec">2</param>
+      <param name="example_param_read_for_sec">2</param>
+    </hardware>
+    <joint name="joint1">
+      <command_interface name="position">
+        <param name="min">-1</param>
+        <param name="max">1</param>
+      </command_interface>
+      <state_interface name="position"/>
+    </joint>
+    <joint name="joint2">
+      <command_interface name="position">
+        <param name="min">-1</param>
+        <param name="max">1</param>
+      </command_interface>
+      <state_interface name="position"/>
+    </joint>
+    <gpio name="flange_analog_IOs">
+      <command_interface name="analog_output1"/>
+      <state_interface name="analog_output1"/> <!-- Needed to know current state of the output -->
+      <state_interface name="analog_input1"/>
+      <state_interface name="analog_input2"/>
+    </gpio>
+    <gpio name="flange_vacuum">
+      <command_interface name="vacuum"/>
+      <state_interface name="vacuum"/> <!-- Needed to know current state of the input -->
+    </gpio>
```

Comment:
Thinking more about this and looking some real-world examples, e.g., [UR robots](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/38a467eb0bc2656e4a7432fcbed2a71d6d762780/ur_robot_driver/src/hardware_interface.cpp#L146-L177), we could/should enable users to define size and the type of the gpio interface.
Those "issues" are discussed in #422, ros-controls/roadmap#45 and multiple times in our meetings.

Therefore I propose the following definition/extensions:

- adding mandatory `type` argument - to support `double`, `bool` and `int` types (needs extension of interfaces)
- adding optional `size` argument - the numbering [0, size) should be added in the HW and controller
- defining inputs by using `state` interfaces only
- defining output by using  `command` and `state` interfaces

```suggestion
   <gpio name="flange_vacuum" type="bool">
      <command_interface name="vacuum"/>
      <state_interface name="vacuum"/>
    </gpio>
    <gpio name="flange" type="bool" size="8">
      <command_interface name="digital_output"/>
      <state_interface name="digital_output"/>
    </gpio>
    <gpio name="flange" type="double" size="2">
      <state_interface name="analog_input"/>
    </gpio>
```

Having this described in URDF is the first step. This description could be optimized when developing an abstraction using semantic components, controller or using it inside hardware.

Or do we need something more generic? What?

@Karsten1987 @bmagyar @olivier-stasse @jordan-palacios @v-lopez @AugusteBourgois @ejalaa12 @mahaarbo

Please comment if you have any opinion on this.

Reply:
@destogl no strong opinion on this on my side. What you are proposing here looks generic enough for me. Just one small note: @mahaarbo uses the ros2_control_demos for demonstration and this is nice.
Would it be possible to merge [PR 101 from ros2_control_demos](https://github.com/ros-controls/ros2_control_demos/pull/101) on branch add-rrbot-extensions. This would makes your [PR 76](https://github.com/ros-controls/ros2_control_demos/pull/76) able to be merged in master of ros2_control_demos.
The structure you (@destogl) propose in this PR is nice and would be very much appreciated for all the new examples.
@mahaarbo thanks for the nice extension.

Reply:
If I understand correctly, you mean that 
```xml
<gpio name="flange" type="bool" size="8">
  <command_interface name="digital_output"/>
  <state_interface name="digital_output"/>
</gpio>
```
would be a helpful tag to fill the `hardware_info` just as though we had written all 8 of them out, and then we have some other helper function to create the 8 command and 8 state interfaces needed in the `hardware_interface` and in the controllers, but the underlying handles are still only a single `bool`? Or did you mean that the handle has a pointer to an array of `bool`s?

Reply:
I would go with single `bool` for now in handles. Adding support for arrays we can consider in the future.

Reply:
I quite like this approach, also in favour of the array extension!